### PR TITLE
fix(web): late callback dedup across invocation boundaries

### PR DESCRIPTION
**Pull request by mindfn**
## Summary
- Preserve just-finalized foreground/background stream refs across one invocation boundary so true late callbacks can still replace their original stream bubble
- Fence finalized-fallback replacement with time/content and reply-target compatibility guards so callbacks do not overwrite a different invocation or reply chain
- Disable store-level callback→finalized-stream bridging without invocation evidence, and mark callback-only unmatched bubbles so the store does not re-merge them heuristically
- Restore late-stream suppression for callback-first, invocationless-placeholder, and finalized-fallback flows in both active and background paths

## Scope
- Foreground hook: `packages/web/src/hooks/useAgentMessages.ts`
- Background flow: `packages/web/src/hooks/useSocket-background.ts`, `packages/web/src/hooks/useSocket-background-system-info.ts`, `packages/web/src/hooks/useSocket.ts`, `packages/web/src/hooks/useSocket-background.types.ts`
- Reply-target compatibility: `packages/web/src/hooks/reply-target-compat.ts`
- Store dedup contract: `packages/web/src/stores/chatStore.ts`, `packages/web/src/stores/chat-types.ts`
- Regression coverage: `packages/web/src/hooks/__tests__/useAgentMessages-late-callback-dedup.test.ts`, `packages/web/src/hooks/__tests__/useSocket-background.test.ts`, `packages/web/src/stores/__tests__/chatStore-multithread.test.ts`

## Related
- Follow-up to #266 (`fix(hub): duplicate message rendering in console`)
- Protocol/backend follow-up: #454 (`always propagate callback invocationId to avoid late-callback duplicate bubbles`)
- Code/tests still contain internal `#586` / `issue-586-state-trifecta` labels, but this repo does not have a GitHub issue `#586`; this PR does not close one

## Verification
- `pnpm check`
- `pnpm --filter @cat-cafe/web exec vitest run src/hooks/__tests__/useAgentMessages-late-callback-dedup.test.ts src/hooks/__tests__/useSocket-background.test.ts src/stores/__tests__/chatStore-multithread.test.ts`
- Additional follow-up regressions were added in the touched test files during review iterations; the exact passing count changed across later heads, so the commands above are the source of truth

### CODE REVIEW

**Comment by mindfn** [comment]
@codex review

Rebased onto latest `main`, kept the late-callback fence fix, and pushed `8f4204a5` to address the merge-commit lint failure (`project-setup-card-ime.test.tsx` formatting). Local verification:
- `pnpm check`
- `pnpm --filter @cat-cafe/web exec vitest run src/hooks/__tests__/useAgentMessages-late-callback-dedup.test.ts src/hooks/__tests__/useSocket-background.test.ts src/stores/__tests__/chatStore-multithread.test.ts`

**Comment by mindfn** [comment]
@codex review

Addressed the two new P1s on top of `5c67618c`.

What changed:
- Unmatched callback-only bubbles no longer inherit an inferred `invocationId`; only explicit callback invocation IDs are persisted.
- Added an internal `callbackBridge.skipDedup` guard so store Phase 2/3 dedup will not re-bridge a callback bubble that the hook layer already proved was unmatched.
- Added regression coverage for both active-thread and background-thread paths.

Local verification:
- `pnpm check`
- `pnpm --filter @cat-cafe/web exec vitest run src/hooks/__tests__/useAgentMessages-late-callback-dedup.test.ts src/hooks/__tests__/useSocket-background.test.ts src/stores/__tests__/chatStore-multithread.test.ts`

**Comment by mindfn** [comment]
@codex review

Follow-up on `ad2f6b17` for the new P2s.

Adjustment:
- Restored late-stream suppression for callback-first flows without explicit `invocationId`.
- Kept the P1 safety fence: inferred suppression now only kicks in when there is no dangling finalized ref for that cat/thread, so delayed callbacks from an older invocation still stay isolated.
- Added callback-first regression coverage for both active and background paths.

Local verification:
- `pnpm check`
- `pnpm --filter @cat-cafe/web exec vitest run src/hooks/__tests__/useAgentMessages-late-callback-dedup.test.ts src/hooks/__tests__/useSocket-background.test.ts src/stores/__tests__/chatStore-multithread.test.ts`

**Comment by mindfn** [comment]
@codex review

Addressed the latest P1 on `f4c01ad3`.

Local verification:
- `pnpm check`
- `pnpm --filter @cat-cafe/web exec vitest run src/hooks/__tests__/useAgentMessages-late-callback-dedup.test.ts src/hooks/__tests__/useSocket-background.test.ts src/stores/__tests__/chatStore-multithread.test.ts`

This patch prevents fenced fallback callback merges from writing an inferred replacement lock for the new invocation, while preserving callback-first late-stream suppression when the replacement matched by invocation ID.

**Comment by mindfn** [comment]
@codex review

Addressed the latest P2s on `13fb8eff`.

Local verification:
- `pnpm check`
- `pnpm --filter @cat-cafe/web exec vitest run src/hooks/__tests__/useAgentMessages-late-callback-dedup.test.ts src/hooks/__tests__/useSocket-background.test.ts src/stores/__tests__/chatStore-multithread.test.ts`

This keeps inferred suppression only for exact matches or active invocationless placeholders, while fenced finalized fallback merges still avoid locking the new invocation.

**Comment by mindfn** [comment]
@codex review

Addressed the latest P2 on `f0b17dc8`.

Local verification:
- `pnpm check`
- `pnpm --filter @cat-cafe/web exec vitest run src/hooks/__tests__/useAgentMessages-late-callback-dedup.test.ts src/hooks/__tests__/useSocket-background.test.ts src/stores/__tests__/chatStore-multithread.test.ts`

Background callback replacement now skips invocationless fallback matching whenever the callback carries an explicit `invocationId`, so explicit callback results can no longer overwrite unrelated placeholder bubbles.

**Comment by mindfn** [comment]
@codex review

Addressed the latest P2s on `859f5524`.

Local verification:
- `pnpm check`
- `pnpm --filter @cat-cafe/web exec vitest run src/hooks/__tests__/useAgentMessages-late-callback-dedup.test.ts src/hooks/__tests__/useSocket-background.test.ts src/stores/__tests__/chatStore-multithread.test.ts`

Callback-only bubbles now keep late-stream suppression even when an older finalized ref exists but was not matchable for the callback. I also applied the same root-cause fix to the background path to keep the two hooks aligned.

**Comment by mindfn** [comment]
@codex review

Addressed the latest P2 on `3c945b39`.

Local verification:
- `pnpm check`
- `pnpm --filter @cat-cafe/web exec vitest run src/hooks/__tests__/useAgentMessages-late-callback-dedup.test.ts src/hooks/__tests__/useSocket-background.test.ts src/stores/__tests__/chatStore-multithread.test.ts`

The store no longer performs a Phase 3 callback→finalized-stream merge without invocation evidence. Hook-level finalized refs remain the only precise late-callback bridge, and ambiguous same-content callbacks now stay separate instead of risking a false cross-invocation merge.

**Comment by mindfn** [comment]
@codex review

Addressed the latest P2s on `2445a9e9`.

What changed:
- Restored late-stream suppression after `finalized_fallback` callback merges by reusing the replaced stream bubble's own `invocationId`, instead of inferring from the current invocation.
- Applied the same fix to the background-thread path.
- Added targeted regression coverage for active/background reordered late-stream chunks after callback-after-done merges.

Local verification:
- `pnpm check`
- `pnpm --filter @cat-cafe/web exec vitest run src/hooks/__tests__/useAgentMessages-late-callback-dedup.test.ts src/hooks/__tests__/useSocket-background.test.ts src/stores/__tests__/chatStore-multithread.test.ts`

**Comment by mindfn** [comment]
@codex review

Rebased onto latest `fix/late-callback-dedup` and pushed `1f66cdb8`.

This keeps late-stream suppression after `finalized_fallback` callback merges by reusing the replaced stream bubble's own `invocationId` in both active and background paths. Added focused regression coverage for reordered late stream chunks after callback-after-done merges.

Fresh verification after rebase:
- `pnpm check`
- `pnpm --filter @cat-cafe/web exec vitest run src/hooks/__tests__/useAgentMessages-late-callback-dedup.test.ts src/hooks/__tests__/useSocket-background.test.ts src/stores/__tests__/chatStore-multithread.test.ts`

**Comment by mindfn** [comment]
@codex review

Addressed the latest P2s on `ee953837`.

What changed:
- Restored late-stream suppression after `finalized_fallback` callback replacement when the replaced stream bubble finalized before invocation binding and therefore had no stored `invocationId`.
- Kept the prior P1 protection: when the replaced finalized bubble already has its own `invocationId`, we still prefer that over the current inferred invocation and do not relabel the lock.
- Applied the same fix to the background-thread path.
- Added regression coverage for both active/background late-bound finalized-fallback scenarios.

Local verification:
- `pnpm --filter @cat-cafe/web exec vitest run src/hooks/__tests__/useAgentMessages-late-callback-dedup.test.ts src/hooks/__tests__/useSocket-background.test.ts src/stores/__tests__/chatStore-multithread.test.ts`
- `pnpm check`

File: packages/web/src/hooks/useAgentMessages.ts [hunk 1]
```diff
@@ -451,15 +483,26 @@ export function useAgentMessages() {
             activeRefs.current.delete(msg.catId);
             // Consume the finalized ref — callback successfully replaced the bubble
             finalizedStreamRef.current.delete(msg.catId);
-            if (invocationId) {
-              replacedInvocationsRef.current.set(msg.catId, invocationId);
+            if (explicitInvocationId) {
+              replacedInvocationsRef.current.set(msg.catId, explicitInvocationId);
+            } else if (replacementTarget?.matchKind === 'finalized_fallback') {
+              if (replacementTarget.replacementInvocationId) {
```
[the remark this answers is not in the document]
Comment (mindfn) [reply packages/web/src/hooks/useAgentMessages.ts]: Fixed in `ee953837`.

The active-thread path now keeps the `finalized_fallback` suppression lock even when the replaced stream bubble finalized before invocation binding and therefore had no stored `replacementInvocationId`. In that specific late-bound case we fall back to the current inferred invocation; if the finalized bubble already has its own invocation ID, we still prefer that stored ID and do not relabel the lock.

Added regression coverage in `useAgentMessages-late-callback-dedup.test.ts` for the late-bound finalized-fallback scenario, and re-ran:
- `pnpm --filter @cat-cafe/web exec vitest run src/hooks/__tests__/useAgentMessages-late-callback-dedup.test.ts src/hooks/__tests__/useSocket-background.test.ts src/stores/__tests__/chatStore-multithread.test.ts`
- `pnpm check`

File: packages/web/src/hooks/useSocket-background.ts [hunk 2]
```diff
@@ -353,22 +380,28 @@ export function handleBackgroundAgentMessage(
...
         // Consume finalized ref — callback successfully replaced
         options.finalizedBgRefs.delete(streamKey);
-        // #586 P1-2 fix: Only set replacedInvocations when we have a real invocationId.
-        // Fallback matches return null — writing a pseudo ID would permanently suppress
-        // future invocationless stream chunks via shouldSuppressLateBackgroundStreamChunk.
-        if (replacementTarget.invocationId) {
+        if (explicitInvocationId) {
+          options.replacedInvocations.set(streamKey, explicitInvocationId);
+        } else if (replacementTarget.invocationId) {
+          // finalized_fallback must reuse the replaced stream bubble's own
```
[the remark this answers is not in the document]
Comment (mindfn) [reply packages/web/src/hooks/useSocket-background.ts]: Fixed in `ee953837`.

The background path now mirrors the active-thread fix: when a `finalized_fallback` target has no stored invocation ID because the stream finalized before binding, we still propagate a suppression lock using the late-bound thread invocation. If the replaced finalized bubble already carries its own invocation ID, that stored ID still wins so we do not regress the earlier cross-invocation fence fix.

Added a matching regression in `useSocket-background.test.ts` for the late-bound finalized-fallback case, and re-ran:
- `pnpm --filter @cat-cafe/web exec vitest run src/hooks/__tests__/useAgentMessages-late-callback-dedup.test.ts src/hooks/__tests__/useSocket-background.test.ts src/stores/__tests__/chatStore-multithread.test.ts`
- `pnpm check`

File: packages/web/src/hooks/useAgentMessages.ts [hunk 3]
```diff
@@ -477,12 +523,15 @@ export function useAgentMessages() {
...
-              (!hasExplicitInvocationId || getCurrentInvocationIdForCat(msg.catId) === msg.invocationId);
-            if (shouldLockReplacement) {
-              replacedInvocationsRef.current.set(msg.catId, invocationId);
+            // instead of spawning a second bubble. Explicit callback invocationIds
+            // are always safe. Invocationless callback-only bubbles also need the
+            // inferred lock even when an older finalized ref exists but was not
+            // matchable for this callback; otherwise the current invocation can
+            // still spawn a duplicate late stream bubble.
+            if (explicitInvocationId) {
+              replacedInvocationsRef.current.set(msg.catId, explicitInvocationId);
```
[the remark this answers is not in the document]
Comment (mindfn) [reply packages/web/src/hooks/useAgentMessages.ts]: Fixed in `d0c5d1ac`.

The active-thread unmatched-callback path no longer writes `replacedInvocationsRef` for an explicit callback ID unless the current bound invocation is already known and matches that same ID. That preserves the callback-first suppression behavior for real current invocations, but avoids letting a stale explicit callback suppress live stream chunks while `invocation_created` / `catInvocations` is still catching up.

I also mirrored the same guard in the background-thread path because it had the same stale-suppression hole.

Verification:
- `pnpm --filter @cat-cafe/web exec vitest run src/hooks/__tests__/useAgentMessages-bubble-merge.test.ts src/hooks/__tests__/useAgentMessages-late-callback-dedup.test.ts src/hooks/__tests__/useSocket-background.test.ts src/stores/__tests__/chatStore-multithread.test.ts`
- `pnpm check`

**Comment by mindfn** [comment]
@codex review

Addressed the latest P1 on `d0c5d1ac`.

What changed:
- Unmatched callback bubbles with an explicit `invocationId` now write the late-stream suppression lock only when the current bound invocation already matches that same ID.
- This avoids stale explicit callbacks suppressing legitimate live stream chunks while invocation binding is still catching up.
- Applied the same guard to the background-thread path and added a matching regression test there.

Local verification:
- `pnpm --filter @cat-cafe/web exec vitest run src/hooks/__tests__/useAgentMessages-bubble-merge.test.ts src/hooks/__tests__/useAgentMessages-late-callback-dedup.test.ts src/hooks/__tests__/useSocket-background.test.ts src/stores/__tests__/chatStore-multithread.test.ts`
- `pnpm check`

File: packages/web/src/hooks/useAgentMessages.ts [hunk 4]
```diff
@@ -256,69 +260,94 @@ export function useAgentMessages() {
...
+            m.id === finalizedEntry.messageId && m.type === 'assistant' && m.catId === catId && m.origin === 'stream',
+        );
+        if (finalized) {
+          // #586 P1 guard: if invocation_created has fired since finalization
+          // (fencedAt is set), apply time + content fence to prevent a callback-only
+          // inv-2 from silently overwriting inv-1's finalized bubble.
+          if (finalizedEntry.fencedAt) {
+            const LATE_CALLBACK_GRACE_MS = 5_000;
+            if (Date.now() - finalizedEntry.fencedAt > LATE_CALLBACK_GRACE_MS) return null;
+            if (callbackContent !== undefined && finalized.content !== callbackContent) return null;
```
[the remark this answers is not in the document]
Comment (mindfn) [reply packages/web/src/hooks/useAgentMessages.ts:339]: Fixed in `7bf1bb59`.

The active-thread `finalized_fallback` path now requires `replyTo` compatibility in addition to the existing time/content fence before it can reuse a finalized stream bubble. If the callback targets a different parent, we now leave the finalized bubble untouched and create a separate callback bubble instead of rewriting the old reply chain.

Added a regression in `useAgentMessages-late-callback-dedup.test.ts` for the same-content/different-replyTo callback-after-done case, and re-ran:
- `pnpm --filter @cat-cafe/web exec vitest run src/hooks/__tests__/useAgentMessages-bubble-merge.test.ts src/hooks/__tests__/useAgentMessages-late-callback-dedup.test.ts src/hooks/__tests__/useSocket-background.test.ts src/stores/__tests__/chatStore-multithread.test.ts`
- `pnpm check`

File: packages/web/src/hooks/useSocket-background.ts [hunk 5]
```diff
@@ -185,20 +199,32 @@ function findBackgroundCallbackReplacementTarget(
...
+        m.id === finalizedEntry.messageId && m.type === 'assistant' && m.catId === msg.catId && m.origin === 'stream',
     );
     if (finalized) {
-      return { id: finalized.id, invocationId: invocationId ?? null };
+      // #586 P1 guard: time + content fence across invocation boundary
+      if (finalizedEntry.fencedAt) {
+        const LATE_CALLBACK_GRACE_MS = 5_000;
+        if (Date.now() - finalizedEntry.fencedAt > LATE_CALLBACK_GRACE_MS) return null;
+        if (msg.content !== undefined && finalized.content !== undefined && finalized.content !== msg.content)
+          return null;
```
[the remark this answers is not in the document]
Comment (mindfn) [reply packages/web/src/hooks/useSocket-background.ts:223]: Fixed in `7bf1bb59`.

The background-thread `finalized_fallback` path now applies the same `replyTo` fence as the foreground hook. Same-content callbacks only reuse the finalized bubble when they stay on the same reply chain; otherwise we keep the old finalized message and add a new callback bubble for the new parent.

Added a matching regression in `useSocket-background.test.ts`, and re-ran:
- `pnpm --filter @cat-cafe/web exec vitest run src/hooks/__tests__/useAgentMessages-bubble-merge.test.ts src/hooks/__tests__/useAgentMessages-late-callback-dedup.test.ts src/hooks/__tests__/useSocket-background.test.ts src/stores/__tests__/chatStore-multithread.test.ts`
- `pnpm check`

**Comment by mindfn** [comment]
@codex review

Addressed the latest P2s on `7bf1bb59`.

What changed:
- `finalized_fallback` now requires `replyTo` compatibility in addition to the existing time/content fence before it can reuse a finalized stream bubble.
- Applied the same reply-target fence to the background-thread path.
- Added active/background regressions for same-content callbacks with different reply chains after `done` + `invocation_created`.

Local verification:
- `pnpm --filter @cat-cafe/web exec vitest run src/hooks/__tests__/useAgentMessages-bubble-merge.test.ts src/hooks/__tests__/useAgentMessages-late-callback-dedup.test.ts src/hooks/__tests__/useSocket-background.test.ts src/stores/__tests__/chatStore-multithread.test.ts`
- `pnpm check`

File: packages/web/src/hooks/reply-target-compat.ts [hunk 6]
```diff
@@ -0,0 +1,3 @@
+export function isReplyTargetCompatible(existingReplyTo?: string, incomingReplyTo?: string): boolean {
+  return (existingReplyTo ?? null) === (incomingReplyTo ?? null);
```
[the remark this answers is not in the document]
Comment (mindfn) [reply packages/web/src/hooks/reply-target-compat.ts]: Fixed in `ed28553a`.

I relaxed `isReplyTargetCompatible` so an unknown reply target on the finalized stream bubble is treated as "missing evidence", not a hard mismatch. That preserves the previous P2 fence for concrete reply-target mismatches, but allows legitimate finalized-fallback merges when the original stream placeholder never captured `replyTo` and the late callback is the first event that brings reply metadata.

Added active/background regressions for the exact case the review describes (finalized stream without `replyTo`, late callback with concrete `replyTo`), and re-ran:
- `pnpm --filter @cat-cafe/web exec vitest run src/hooks/__tests__/useAgentMessages-bubble-merge.test.ts src/hooks/__tests__/useAgentMessages-late-callback-dedup.test.ts src/hooks/__tests__/useSocket-background.test.ts src/stores/__tests__/chatStore-multithread.test.ts`
- `pnpm check`

**Comment by mindfn** [comment]
@codex review

Addressed the latest P2 on `ed28553a`.

What changed:
- `isReplyTargetCompatible` now treats a missing `replyTo` on the finalized stream bubble as unknown evidence rather than a hard mismatch.
- Concrete reply-target mismatches still block `finalized_fallback`, so the previous reply-chain fence remains intact.
- Added active/background regressions for the late-callback case where the finalized stream placeholder had no reply metadata but the callback does.

Local verification:
- `pnpm --filter @cat-cafe/web exec vitest run src/hooks/__tests__/useAgentMessages-bubble-merge.test.ts src/hooks/__tests__/useAgentMessages-late-callback-dedup.test.ts src/hooks/__tests__/useSocket-background.test.ts src/stores/__tests__/chatStore-multithread.test.ts`
- `pnpm check`

File: packages/web/src/hooks/reply-target-compat.ts [hunk 7]
```diff
@@ -0,0 +1,6 @@
+export function isReplyTargetCompatible(existingReplyTo?: string, incomingReplyTo?: string): boolean {
+  if (existingReplyTo == null) {
+    return true;
+  }
+  return existingReplyTo === (incomingReplyTo ?? null);
```
[the remark this answers is not in the document]
Comment (mindfn) [reply packages/web/src/hooks/reply-target-compat.ts]: Fixed in `f9bcc669`.

I made `isReplyTargetCompatible` symmetric: it now treats missing `replyTo` on either side as unknown evidence and only blocks `finalized_fallback` when both reply targets are present and different. That keeps the previous concrete-mismatch fence, while restoring legitimate callback-after-done merges when either the finalized stream bubble or the callback omitted reply metadata.

Added active/background regressions for the missing-callback-reply case, and re-ran:
- `pnpm --filter @cat-cafe/web exec vitest run src/hooks/__tests__/useAgentMessages-bubble-merge.test.ts src/hooks/__tests__/useAgentMessages-late-callback-dedup.test.ts src/hooks/__tests__/useSocket-background.test.ts src/stores/__tests__/chatStore-multithread.test.ts`
- `pnpm check`

**Comment by mindfn** [comment]
@codex review

Addressed the latest P2 on `f9bcc669`.

What changed:
- `isReplyTargetCompatible` now only rejects `finalized_fallback` when both reply targets are present and different.
- Missing `replyTo` on either the finalized stream bubble or the callback is treated as unknown evidence, so legitimate late callback merges still go through.
- Added active/background regressions for the missing-callback-reply case.

Local verification:
- `pnpm --filter @cat-cafe/web exec vitest run src/hooks/__tests__/useAgentMessages-bubble-merge.test.ts src/hooks/__tests__/useAgentMessages-late-callback-dedup.test.ts src/hooks/__tests__/useSocket-background.test.ts src/stores/__tests__/chatStore-multithread.test.ts`
- `pnpm check`

File: packages/web/src/hooks/useAgentMessages.ts [hunk 8]
```diff
@@ -451,15 +486,29 @@ export function useAgentMessages() {
             activeRefs.current.delete(msg.catId);
             // Consume the finalized ref — callback successfully replaced the bubble
             finalizedStreamRef.current.delete(msg.catId);
-            if (invocationId) {
-              replacedInvocationsRef.current.set(msg.catId, invocationId);
+            if (explicitInvocationId) {
+              replacedInvocationsRef.current.set(msg.catId, explicitInvocationId);
+            } else if (replacementTarget?.matchKind === 'finalized_fallback') {
+              const suppressionInvocationId = replacementTarget.replacementInvocationId ?? inferredInvocationId;
+              if (suppressionInvocationId) {
```
[the remark this answers is not in the document]
Comment (mindfn) [reply packages/web/src/hooks/useAgentMessages.ts]: Fixed in `6914bb74`. `finalized_fallback` now carries the fence state through the replacement target, and the suppression lock only borrows the current invocation for unfenced late-bind fallback. Once `invocation_created` fenced the finalized ref, we only suppress against the replaced bubble's own invocation ID; if that bubble never captured one, we leave suppression unset so the new invocation's first stream chunk is not dropped. Added a regression in `useAgentMessages-late-callback-dedup.test.ts` and reran the targeted suite, the 4-file regression set, and `pnpm check`.

File: packages/web/src/hooks/useSocket-background.ts [hunk 9]
```diff
@@ -353,22 +382,34 @@ export function handleBackgroundAgentMessage(
...
-        // Fallback matches return null — writing a pseudo ID would permanently suppress
-        // future invocationless stream chunks via shouldSuppressLateBackgroundStreamChunk.
-        if (replacementTarget.invocationId) {
-          options.replacedInvocations.set(streamKey, replacementTarget.invocationId);
+        if (explicitInvocationId) {
+          options.replacedInvocations.set(streamKey, explicitInvocationId);
+        } else {
+          const suppressionInvocationId =
+            replacementTarget.invocationId ??
+            (replacementTarget.matchKind === 'finalized_fallback' ? getThreadInvocationId(msg, options) : undefined);
```
[the remark this answers is not in the document]
Comment (mindfn) [reply packages/web/src/hooks/useSocket-background.ts]: Fixed in `6914bb74`. Background `finalized_fallback` now propagates whether the finalized ref was fenced by `invocation_created`, and only unfenced late-bind fallback may reuse the current thread invocation for suppression. Fenced fallback without a stored invocation ID no longer locks suppression onto the new invocation, so legitimate background stream chunks survive. Added a matching regression in `useSocket-background.test.ts` and reran the targeted suite, the related 4-file regression set, and `pnpm check`.

**Comment by mindfn** [comment]
@codex review

Latest fixes are in `6914bb74`.

File: packages/web/src/hooks/useSocket-background.ts [hunk 10]
```diff
@@ -163,11 +175,15 @@ function findBackgroundCallbackReplacementTarget(
...
       ) {
-        return { id: m.id, invocationId };
+        return { id: m.id, invocationId, matchKind: 'exact' };
       }
     }
   }
 
+  if (explicitInvocationId) {
+    return null;
+  }
```
[the remark this answers is not in the document]
Comment (mindfn) [reply packages/web/src/hooks/useSocket-background.ts:185]: I checked this one locally and I’m pushing back on the suggested relaxation as written.

I changed the background matcher to let explicit callbacks fall through to `active_invocationless`, then reran `src/hooks/__tests__/useSocket-background.test.ts`. That immediately reintroduced the existing P1 regression `stale explicit callback bubble must NOT suppress live background stream chunks`: a stale explicit callback from the previous invocation now overwrites the live invocationless placeholder and the real stream never recovers.

The issue is that in the background path an active invocationless placeholder with no bound thread invocation does not provide enough evidence to distinguish:
- the intended "same invocation, binding arrived late" callback
- a stale explicit callback from the previous invocation while the new stream is still unbound

Those two states are observationally equivalent here. Allowing the fallback turns that ambiguity into destructive overwrite/suppression, which is worse than the current duplicate-callback behavior.

So I’m keeping the current rule: explicit callbacks may exact-match a bound invocation, but they do not fall back to invocationless placeholders. If we want to support this race safely, we need a stronger correlation signal than "active invocationless placeholder" (for example, backfilling a trustworthy invocation ID onto the early background bubble before callback reconciliation).

**Comment by zts212653** [comment]
Hi @mindfn — thanks for the deep investigation and the thorough regression coverage. The race condition you've identified is real: clearing `finalizedStreamRef` immediately on `invocation_created` leaves a window where late callbacks from the previous invocation can't find their replacement target, creating duplicate bubbles.

That said, we think the fix can be approached more cleanly at a different layer.

### Why we're hesitant about the current approach

The PR adds four new coordination mechanisms (`fencedAt`, `matchKind`, `isReplyTargetCompatible`, `callbackBridge.skipDedup`) across both active and background paths. Each layer addresses a real edge case, but the cumulative complexity is high — the Codex reviews surfaced 15+ P1/P2 findings, many of which were edge cases introduced by the new guards themselves. This is the fundamental challenge with client-side heuristic stacking: each fix narrows one gap but can open others.

### What we think is a better path

**Option A — Protocol-level fix (preferred):**

The backend callback routes already carry `invocationId` in most broadcast payloads. The root cause of the `invocationless` fallback complexity is that some callback paths don't propagate it to the WebSocket event. If we close that gap at the API layer — ensuring every `origin: 'callback'` message always carries `invocationId` — the `findCallbackReplacementTarget` exact-match path works universally, and the entire `findInvocationlessStreamPlaceholder` / `finalized_fallback` / fencing machinery becomes unnecessary.

This is a one-time backend change that eliminates a whole class of client-side race conditions rather than fencing them case by case.

**Option B — Lightweight client-side alternative (if backend change is deferred):**

Instead of fencing `finalizedStreamRef` with `fencedAt` + content/reply-target guards, use a simple TTL map on stream finalization:

```ts
// On done(isFinal): record invocationId → messageId with auto-expiry
const LATE_CALLBACK_GRACE_MS = 5_000;
finalizedStreamRef.current.set(catId, { messageId, invocationId });
setTimeout(() => {
  const entry = finalizedStreamRef.current.get(catId);
  if (entry?.messageId === messageId) finalizedStreamRef.current.delete(catId);
}, LATE_CALLBACK_GRACE_MS);
```

`invocation_created` no longer needs to touch `finalizedStreamRef` at all — the TTL handles cleanup. Late callbacks within the window find the ref; after the window, the ref is gone. No fencing state, no `matchKind` discrimination, no content/reply-target guards needed.

### What we'd love to keep

Your regression test cases are excellent — they precisely describe real race conditions (`late callback with SAME/DIFFERENT content across boundary`, `fenced finalized_fallback`, `reply-target mismatch`, etc.). We'd like to absorb those test scenarios regardless of which implementation path we take.

### Suggested next step

Could you open a dedicated issue describing the specific race condition (late callback after `invocation_created` clears `finalizedStreamRef`)? The original #266 is closed, and having a focused accepted issue would let us track this properly. We're happy to triage it immediately.

Thanks again for the thorough work — this is clearly a well-understood problem, and we appreciate the contribution. 🐾

**Comment by mindfn** [comment]
Thanks — I agree the protocol-level fix is the cleaner long-term direction.

I opened a dedicated issue to track that root cause and invariant here: #454.

The main reason this PR went down the client-side path was to stop the user-visible duplicate-bubble race without waiting on a backend contract change. Regardless of which implementation path we choose, I think the regression scenarios captured here are still valuable and should be preserved.

At this point I see two reasonable options:
- Land this PR as a bounded client-side mitigation, then follow up in #454 with the protocol/backend cleanup that removes the remaining heuristic surface.
- Close this PR and move the fix to #454 if you'd prefer to avoid merging the client-side coordination state.

I'm fine with either direction; the key thing is that we now have a focused issue that tracks the actual race condition separately from #266.

**Comment by zts212653** [comment]
Thanks for the clean follow-up and for opening #454 — that's exactly the right anchor.

**Our decision: we'll pursue the protocol-level fix via #454 and won't merge this PR.**

Reasoning:

- Since we both agree the backend invariant (`invocationId` on every callback payload) is the cleaner long-term answer, merging 2400 lines of client-side coordination state into `main` as a temporary bridge doesn't pay for itself — especially when the heuristic surface has already shown it generates new edge cases during review.
- The regression test scenarios you wrote here are genuinely valuable. We'll preserve those cases (same/different content across boundary, reply-target mismatch, callback-first ordering, background-thread variants) as red-light fixtures when we implement #454.

We've triaged #454 as `bug` + `accepted`. If you'd like to take a crack at the backend side as well, we're happy to review — the callback broadcast paths in `callbacks.ts` already carry `invocationId` internally, so the delta should be small.

Thanks again for the thorough analysis — the issue quality and willingness to pivot to the right fix layer is exactly the kind of contribution we value. 🐾

**Comment by mindfn** [comment]
Understood. I won't push the client-side mitigation on this branch any further.

I'll treat #410 as superseded by #454, and the regression scenarios from this PR can be carried over there as red-light fixtures for the protocol-level implementation.

Closing this PR so the tracking state matches the agreed direction.
